### DISCUSSION TITLE
I want to unregistered the confirm password field after validation

### DISCUSSION THREAD

**Discussion by azad-steadybiz** [react-hook-form/react-hook-form#2205] [Q&A] [answered]
I want to unregistered the confirm password field from data object after clicking on submit button.

**Comment by bluebill1049** [comment]
do in onSubmit function or invoke unreigster

**Reply by azad-steadybiz** [reply]
invoke unregister in onSubmit Function.

**Reply by azad-steadybiz** [reply]
?

**Reply by bluebill1049** [reply]
no, I meant to remove the data in your onSubmit function or unregister the onSubmmit callback.

**Reply by rogix** [reply]
Hi, @bluebill1049 I have a similar question, I think. Would you mind to clarify more the answer?

I am not sure if unregister is what I really want here, but the Idea is the same as the question: 

Let's say I have a form with `Password` and `ConfirmPassword` but I don't actually need the `ConfirmPassord` data, it is only needed to check is the passwords matches and activate the button `Submit` for example. So, once the check is done `ConfirmPassword` should be "ignored" and not sent to `data`. Is there a way to _ignore_ a field? I couldn't find it in the docs. What do you suggest? Thanks!

**Reply by bluebill1049** [reply]
perhaps you don't want to register that `confirmPassowrd`? keep the state inside a `useState` and all you want is to confirm both password are actually equal with `validate` function.

**Reply by rogix** [reply]
Thanks, @bluebill1049! Wouldn't lose the use form validation doing this  way?

**Comment by Moshyfawn** [comment, marked as answer]
Maybe I'm missing something, but.. wouldn't this be the easiest solution _(and the most compliant with the Web API)_:

- Have both `password` and `confirmPassword` fields registered
- Validate both fields against each other, what I assume, would be `onChange` as you want to control the submit button's `disabled` state
- In submit handler just.. take the `data.password` value and pass into your API method; simply ignore the `confirmPassword`value

**Reply by bluebill1049** [reply]
I would vote for both, unless they really have an issue with `confirmPassword` as a form value, but go back to the original comment, you can easily omit that value in the onSubmit callback.

**Reply by Moshyfawn** [reply]
I can't find a use-case where a form value you have in your form would be so "unwanted" in your submit handler, TBH. Even the `onSubmit` validation happens before you get your data in the handler.

Having a plain HTML form, you'd always get all your form field's values and then transform it accordingly to your API method's payload.

**Reply by rogix** [reply]
Hi, @Moshyfawn! Yeah, it's a very specific use case where I have multiple forms and I need to get data from all of them in a "wizard-tunnel" kind of form. The point is that I was creating a single object and using _MyObjectType_ (`useForm<MyObjectType>`) to all the forms and filling it with the data gathered from each form till the last one. 

I could do it like you said and only send the password, but frankly i am not very proficient in TS and I get this warning because the others types are missing... anyway. I was just looking for simple way to "isolate" this specific value, so I wouldn't need to specify the same "MyObjectType" to the form and deal with this case. So my issue here is with TS actually. 

Thanks for the insight, though.

**Reply by Moshyfawn** [reply]
Hey, @rogix!

I see. I don't know your specific "wizard-tunnel" implementation, but I'd still **keep all the partial form values** wherever you store it in-between the form steps _(`localStorage`, React `state` etc.)_ and simply send the actual contract values as a payload.

That way you can
- Apply partial types from the main form values type _(you've mentioned the `MyObjectType` type)_ to the individual forms _(steps)_ like
```ts
type AccessCodeStepFormValues = Pick<FormValues, "accessCode">
type PasswordStepFormValues = Pick<FormValues, "password" | "confirmPassword">
```
- Can go back a step having the already submitted values, if need be
- Simply pick the required values to send to your API method collected from each form like
```ts
const onFinalSubmit: SubmitHandler<FormValues> = (data) => {
  const { confirmPassword, accessCode, ...actualApiFields } = data
  
  const payload = actualApiFields
  
  apiClient.send(payload)
}
```

That said, it's, probably, a matter of opinion. I prefer to take the path of least resistance _(at least for me)_ and treat all the steps _(each wizard-tunnel form)_ as a whole form **eventually**. It feels like a closer to the native HTML web form behavior to me.

**Reply by rogix** [reply]
Awesome! 

Thank you very much, @Moshyfawn! That's exactly what I need. 

> That said, it's, probably, a matter of opinion. I prefer to take the path of least resistance (at least for me) and treat all the steps (each wizard-tunnel form) as a whole form eventually. It feels like a closer to the native HTML web form behavior to me.

Yes, that's what I am doing actually. 

Ex:

```javascript
type MyObjectTypes = {
   step1: ObjectTypes1,
   step2: ObjectTypes2,
   ...
}
```
In the end there will be just one object with all the data needed. 

I even thought of doing something like the example of yours, but didn't actually know how to implement it, now I know! This and `partial types`, you've mentioned, will solve my problem very well. Thanks.
